aws/aws.go

```diff
@@ -48,71 +48,84 @@ func AllASGInstanceIds(as []AutoScalingGroup) map[reapable.Region]map[reapable.I
 	return inASG
 }
 
-// returns ASGs as filterables
+// AllAutoScalingGroups describes every AutoScalingGroup in the requested regions
+// *AutoScalingGroups are created for every *autoscaling.AutoScalingGroup
+// and are passed to a channel
 func AllAutoScalingGroups() chan *AutoScalingGroup {
 	ch := make(chan *AutoScalingGroup)
-
-	go func(regions []string) {
-		for _, region := range regions {
+	// waitgroup for all regions
+	wg := sync.WaitGroup{}
+	for _, region := range config.Regions {
+		go func(region string) {
+			// add region to waitgroup
+			wg.Add(1)
 			api := autoscaling.New(&aws.Config{Region: region})
-
-			// TODO: nextToken paging
-			input := &autoscaling.DescribeAutoScalingGroupsInput{}
-			resp, err := api.DescribeAutoScalingGroups(input)
+			err := api.DescribeAutoScalingGroupsPages(&autoscaling.DescribeAutoScalingGroupsInput{}, func(resp *autoscaling.DescribeAutoScalingGroupsOutput, lastPage bool) bool {
+				for _, asg := range resp.AutoScalingGroups {
+					ch <- NewAutoScalingGroup(region, asg)
+				}
+				// if we are at the last page, we should not continue
+				// the return value of this func is "shouldContinue"
+				if lastPage {
+					// on the last page, finish this region
+					wg.Done()
```

Comment:
When the describe API call fails `wg.Done()` is never called and `ch` is never closed (likewise in `AllInstances`).